### PR TITLE
Fix race conditions

### LINKED ISSUE #54: Race condition
When resizing a lot of images one might run into race conditions. See stack trace:

```
ErrorException: fopen(/var/www/vhosts/foobar/httpdocs/releases/106/assets/images/deferred/5/e09d6c21-833fdbd0.jpg.json): failed to open stream: No such file or directory
#13 vendor/contao/image/src/DeferredImageStorageFilesystem.php(87): handleError
#12 src/AppBundle/Sentry/ErrorHandler.php(27): handleError
#11 src/AppBundle/Sentry/ErrorHandler.php(0): fopen
#10 vendor/contao/image/src/DeferredImageStorageFilesystem.php(87): getLocked
#9 vendor/contao/image/src/DeferredResizer.php(87): resizeDeferredImage
#8 vendor/contao/core-bundle/src/Command/ResizeImagesCommand.php(162): resizeImage
#7 vendor/contao/core-bundle/src/Command/ResizeImagesCommand.php(139): execute
#6 vendor/symfony/console/Command/Command.php(255): run
#5 vendor/symfony/console/Application.php(939): doRunCommand
#4 vendor/symfony/framework-bundle/Console/Application.php(87): doRunCommand
#3 vendor/symfony/console/Application.php(273): doRun
#2 vendor/symfony/framework-bundle/Console/Application.php(73): doRun
#1 vendor/symfony/console/Application.php(149): run
#0 vendor/contao/manager-bundle/bin/contao-console(42): null
```

### PR DESCRIPTION
**Pull request by ausi**
Fixes #54 

I found another race condition and fixed it in d6fd14b92ea5a1d3854f25003e71860f6d65b578 but I’m not sure if this works on all platforms.

### CODE REVIEW

**Comment by ausi** [comment]
> but I’m not sure if this works on all platforms.

It doesn’t work on windows: `Failed to remove file "…": unlink(…): Resource temporarily unavailable.`.

I don’t know how to fix that.

**Comment by aschempp** [comment]
Maybe using this hint?
https://github.com/composer/composer/blob/master/src/Composer/Util/Filesystem.php#L194-L200

**Comment by ausi** [comment]
No, that wouldn’t help. I would have to release the lock first before deleting the file and that would enable the race condition again.

**Comment by ausi** [comment]
A possible workaround might be to remove all file contents, release the lock and then unlink the file. When reading the file and it’s empty we could handle it the same way as if it didn’t exist at all.

File: tests/DeferredResizerTest.php [hunk 1]
```diff
@@ -150,6 +150,17 @@ static function (ResizeConfigurationInterface $config, ImageDimensionsInterface
...
 
+        // Calling resizeDeferredImage() a second time should return the already
+        // generated image to prevent race conditions.
+        $resizedImage2 = $resizer->resizeDeferredImage($deferredImage2);
+
+        $this->assertSame($resizedImage->getPath(), $resizedImage2->getPath());
+
+        $this->assertNull(
+            $resizer->resizeDeferredImage($deferredImage2, false),
+            'Non-blocking deffered resize of an existing image should return null'
```
Comment (Toflar) [inline tests/DeferredResizerTest.php]:
```suggestion
            'Non-blocking deferred resize of an existing image should return null'
```

**Comment by ausi** [comment]
As a workaround for windows we now retry to delete the file after releasing the lock if the first delete attemt fails (84f52c9bd552f5b85f09ec9a234445b4a6578bc4)

**Pull request merged**